Upgrade DF to 51

AdamGS opened the pull request:
Testing the upgrade to DF 51.

- https://github.com/apache/datafusion/issues/17558#issuecomment-3487645051

robert3005 commented:
Tihs is blocked on lance and tpchgen-rs because of our benchmark setup

AdamGS committed:
Fix conflicts

Signed-off-by: Adam Gutglick <adam@spiraldb.com>

AdamGS committed:
Patch tpchgen

Signed-off-by: Adam Gutglick <adam@spiraldb.com>

AdamGS commented:
I forgot lance was feature flagged :/

a10y commented:
I ported the schema evolution tests from #5406 and they seem to fail :/

```
aduffy@Andrews-MacBook-Pro /V/C/w/d/vortex-datafusion (adamg/df-51)> cargo test --test schema_evolution
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.12s
     Running tests/schema_evolution.rs (/Volumes/Code/worktree/df51/target/debug/deps/schema_evolution-dc8592b17d16340b)

running 3 tests
test test_filter_schema_evolution_order ... FAILED
test test_filter_with_schema_evolution ... FAILED
test test_filter_schema_evolution_struct_fields ... FAILED

failures:

---- test_filter_schema_evolution_order stdout ----

thread 'test_filter_schema_evolution_order' panicked at vortex-datafusion/tests/schema_evolution.rs:235:10:
called `Result::unwrap()` on an `Err` value: ArrowError(SchemaError("Unable to get field named \"b\". Valid fields: [\"a\"]"), Some(""))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- test_filter_with_schema_evolution stdout ----

thread 'test_filter_with_schema_evolution' panicked at vortex-datafusion/tests/schema_evolution.rs:145:10:
called `Result::unwrap()` on an `Err` value: ArrowError(SchemaError("Unable to get field named \"a\". Valid fields: [\"b\"]"), Some(""))

---- test_filter_schema_evolution_struct_fields stdout ----

thread 'test_filter_schema_evolution_struct_fields' panicked at vortex-datafusion/tests/schema_evolution.rs:409:5:
assertion `left == right` failed
  left: RecordBatch { schema: Schema { fields: [Field { name: "hostname", data_type: Utf8, nullable: true }, Field { name: "payload", data_type: Struct([Field { name: "uptime", data_type: Int64, nullable: true }, Field { name: "instance", data_type: Utf8, nullable: true }]), nullable: true }], metadata: {} }, columns: [StringArray
[
  "host01.local",
  "host01.local",
  "host01.local",
  "host01.local",
  "host02.local",
  "host02.local",
], StructArray
-- validity:
[
  valid,
  valid,
  valid,
  valid,
  valid,
  valid,
]
[
-- child 0: "uptime" (Int64)
PrimitiveArray<Int64>
[
  1,
  2,
  3,
  4,
  10,
  20,
]
-- child 1: "instance" (Utf8)
StringArray
[
  null,
  null,
  null,
  null,
  "c6i",
  "c6i",
]
]], row_count: 6 }
 right: RecordBatch { schema: Schema { fields: [Field { name: "hostname", data_type: Utf8, nullable: true }, Field { name: "payload", data_type: Struct([Field { name: "uptime", data_type: Int64, nullable: true }, Field { name: "instance", data_type: Utf8, nullable: true }]), nullable: true }], metadata: {} }, columns: [StringArray
[
  "host01",
  "host01",
  "host01",
  "host01",
  "host02",
  "host02",
], StructArray
-- validity:
[
  valid,
  valid,
  valid,
  valid,
  valid,
  valid,
]
[
-- child 0: "uptime" (Int64)
PrimitiveArray<Int64>
[
  1,
  2,
  3,
  4,
  10,
  20,
]
-- child 1: "instance" (Utf8)
StringArray
[
  null,
  null,
  null,
  null,
  "c6i",
  "c6i",
]
]], row_count: 6 }


failures:
    test_filter_schema_evolution_order
    test_filter_schema_evolution_struct_fields
    test_filter_with_schema_evolution

test result: FAILED. 0 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.06s

error: test failed, to rerun pass `--test schema_evolution`
aduffy@Andrews-MacBook-Pro /V/C/w/d/vortex-datafusion (adamg/df-51) [101]>
```

a10y committed:
port over cast

Signed-off-by: Andrew Duffy <andrew@a10y.dev>

a10y committed:
cleanup opener

Signed-off-by: Andrew Duffy <andrew@a10y.dev>

a10y committed:
add more

Signed-off-by: Andrew Duffy <andrew@a10y.dev>

a10y committed:
almost there...

Signed-off-by: Andrew Duffy <andrew@a10y.dev>

a10y committed:
modified upstream DF52 schema adapter backport

Signed-off-by: Andrew Duffy <andrew@a10y.dev>

a10y commented:
I pushed a few commits which cover a few fixes:

1. Fix Struct casting to allow for schema-compatible casts on structs (i.e. field reordering, adding nullable columns, dropping columns)
2. Vendor in some changes from last week to the PhysicalExprRewriter that fixes schema evolution for fields of nested struct columns. Relevant block is [here](https://github.com/apache/datafusion/blame/main/datafusion/physical-expr-adapter/src/schema_rewriter.rs#L433-L454)
3. Bring integration tests over from my old PR and get them passing

This does force us to bring in a small block of vendored code that we'll delete when DF 52 releases. We can forego that and just continue to leave the problem open for now, and merge #5717 later. Or we can include the one vendored file which we'll delete on next DF bump.

a10y committed:
remove struct test

Signed-off-by: Andrew Duffy <andrew@a10y.dev>

AdamGS commented:
So @a10y is on vacation for the next few weeks, so just updating my current thinking here:
1. I would love to not vendor temporary fixes (@asubiotto - would love your thoughts here).
2. I'm getting close to unblocking this release but I'm not sure I'll make it this year, if I do - we can release a 51 supporting version (with the underlying issue still there).
3. DF 52 should be [released](https://github.com/apache/datafusion/issues/18566) later this month or in January, which will allow us to quickly release again with the upstream fix.

asubiotto commented:
I'm also on PTO this week so won't be very reachable but I think most of the urgency comes from some changes @brancz is working on so I'll defer to him. Re the vendored code I don't think I mind it too much, especially since it'll be deleted soon. I think the question is whether it's worth the effort to push through this 51 release if 52 is coming soon and most everyone is on holiday starting next week.

robert3005 commented:
My personal bias is to merge the vendored code now since it’s an indication of an upstream bug for a feature we care about. Then delete it once 52 is around. Without it gharchive is broken and it’s a bit annoying to wait for upstream.

brancz commented:
No strong opinion one way or another, we're already at the point where we're carrying several unreleased patches on upstream for both arrow and datafusion.

brancz commented:
Jynxed it, I'm quite certain I'm seeing a very noticeable case for us where we're lacking projection pushdown for a field in a struct (which is one of the things this fixes). Would appreciate it if we could merge this.

AdamGS commented:
This is currently blocked by #5676, once I merge that (hopefully today), I'll rebase this PR and try and catch up on all the pieces here and get it merged.

AdamGS committed:
Dependencies and clippy touch ups

Signed-off-by: Adam Gutglick <adam@spiraldb.com>

AdamGS commented:
I've rebased this PR and fixed a couple of small random issues that came up, the codspeed regressions all point back to Arrow, but that might be an artifact of the runtime because the only relevant changes in Arrow have some pretty convincing benchmarks? I'm not worried about them.

AdamGS commented:
I can't reproduce this TPC-DS issue, if it keeps popping up I'll dig deeper into it.

AdamGS commented:
@asubiotto if you want to take look here, @a10y is OOO until the 7th IMO, not sure if we have anyone else that can weigh in.

AdamGS commented:
finally managed to get a repro of the TPC-DS issue, it has something to do with the new decimal types, I think we should fix that before we merge this, so I'll keep working on it.
The rest of this PR shouldn't change, so would appreciate more eyes if anyone has the cycles.

AdamGS commented:
Fixing TPC-DS here might require a deeper fix in DF, for now I'll remove some of the DF/Decimal code paths

AdamGS committed:
Disable small decimal to-arrow conversion

Signed-off-by: Adam Gutglick <adam@spiraldb.com>

AdamGS commented on vortex-dtype/src/arrow.rs:
```diff
@@ -214,7 +214,7 @@ impl DType {
                 let scale = dt.scale();
 
                 match precision {
-                    // This code is commented out until DataFusion improves its support for smaller decimals.
```
This will stay for now, I've hit an overflow in TPC-DS q1, and I might take a bigger stab at how to handle it in DF.
